refactor(trie): return `OpProofsStorageError` from `execute_and_store_block_updates`

Himess opened the pull request:
Closes #523

Himess committed:
refactor(trie): return OpProofsStorageError from execute_and_store_block_updates

Replace eyre::Result with Result<(), OpProofsStorageError> in the
execute_and_store_block_updates method to improve type safety and
limit eyre usage to CLI crates.

Changes:
- Add ExecutionError and ProviderError variants to OpProofsStorageError
- Implement From traits for BlockExecutionError and ProviderError
- Update method signature to return specific error type
- Remove unnecessary .into() conversions

Closes #523

Himess committed:
fix: store error messages as String to satisfy Clone bound

Himess committed:
fix: update tests to use OpProofsStorageError directly

Since execute_and_store_block_updates now returns Result<(), OpProofsStorageError>
instead of eyre::Result<()>, we no longer need downcast_ref to access the error.

emhane commented on crates/optimism/trie/src/error.rs:
```diff
@@ -82,6 +84,24 @@ pub enum OpProofsStorageError {
     TryLockError,
+    /// Error occurred during block execution.
+    #[error("block execution error: {0}")]
+    ExecutionError(String),
```
```suggestion
    #[error(transparent)]
    ExecutionError(#[from] BlockExecutionError),
```

emhane reviewed (commented):
we don't want any variants to wrap `String` because it's not type safe to identify errors in tests for example based on some error string

Himess committed:
fix: use Arc to wrap non-Clone error types

Use Arc<BlockExecutionError> and Arc<ProviderError> instead of String
to maintain type safety while satisfying the Clone bound on OpProofsStorageError.

This addresses the reviewer feedback about not using String for error variants.

Himess committed:
refactor: use #[from] directly and remove Clone derive

Himess committed:
Merge branch 'unstable' into fix/execute-store-error-type

Himess committed:
test: fix roundtrip test after Clone removal

emhane committed:
Merge branch 'unstable' into fix/execute-store-error-type

emhane reviewed (changes requested):
pls revert changes to conversion from `DatabaseError`. wrap errors in variants in `Arc` if needed.

Himess committed:
refactor(trie): wrap errors in Arc and restore DatabaseError conversion

- Restore Clone derive on OpProofsStorageError
- Wrap BlockExecutionError and ProviderError in Arc for Clone compatibility
- Restore original From<DatabaseError> conversion that recovers the original type

Himess commented:
@emhane Done

Himess committed:
Merge branch 'unstable' into fix/execute-store-error-type

emhane reviewed (approved):
nice

emhane committed:
Merge branch 'unstable' into fix/execute-store-error-type
